File: python/day_19b.py
#!/usr/bin/python3

# No need to change the input data
update_rules = dict()
update_rules[8] = "42 | 42 8"
update_rules[11] = "42 31 | 42 11 31"

known_rules = dict()
subsections_of_rules = dict()
previously_seen_pairs = dict()

def DoesMsgFollowRule(message, source):
    if source in known_rules:
        if known_rules[source] == message:
            return True
        else:
            return False
    if (message, source) in previously_seen_pairs:
        return previously_seen_pairs[(message, source)]
    for subsection in subsections_of_rules[source]:
        parts = len(subsection)
        if parts == 1:
            if(DoesMsgFollowRule(message, subsection[0])):
                previously_seen_pairs[(message, source)] = True
                return True;
        elif parts == 2:
            for i in range(0, len(message) - 1):
                if DoesMsgFollowRule(message[0:i + 1], subsection[0]) and DoesMsgFollowRule(message[i + 1:len(message)], subsection[1]):
                    previously_seen_pairs[(message, source)] = True
                    return True;
        elif parts == 3:
            for i in range(0, len(message) - 2):
                for j in range(i+1, len(message) - 1):
                    if DoesMsgFollowRule(message[0 : i + 1], subsection[0]) and DoesMsgFollowRule(message[i + 1 : j + 1], subsection[1]) and DoesMsgFollowRule(message[j + 1 : len(message)], subsection[2]):
                        previously_seen_pairs[(message, source)] = True
                        return True;
    previously_seen_pairs[(message, source)] = False
    return False


def main():
    file = open("../input/day_19_input")
    messages = list()
    for line in file:
        line = line.strip('\n')
        if line == "":
            continue
        index = line.find(':')
        if index != -1:
            rule_lhs = int(line[0:index])
            rule_rhs = line[index + 1:len(line)]
            if rule_lhs in update_rules:
                rule_rhs = update_rules[rule_lhs]
            if rule_lhs in known_rules:
                rule_rhs = known_rules[lhs]
            index = rule_rhs.find("\"")
            if index != -1:
                known_rules[rule_lhs] = rule_rhs[index + 1]
            else:
                rule_subsection = list()
                num_in_rule_section = 0
                for c in rule_rhs:
                    if c == ' ':
                        if num_in_rule_section > 0:
                            rule_subsection.append(num_in_rule_section)
                        num_in_rule_section = 0
                    elif c == '|':
                        if rule_lhs not in subsections_of_rules:
                            subsections_of_rules[rule_lhs] = list()
                        subsections_of_rules[rule_lhs].append(rule_subsection)
                        rule_subsection = list()
                    else:
                        num_in_rule_section = 10 * num_in_rule_section + int(c)
                rule_subsection.append(num_in_rule_section)
                if rule_lhs not in subsections_of_rules:
                    subsections_of_rules[rule_lhs] = list()
                subsections_of_rules[rule_lhs].append(rule_subsection);
        else:
            messages.append(line)

    res = 0;
    for message in messages:
        if DoesMsgFollowRule(message, 0):
           res += 1
        previously_seen_pairs.clear();
    print(res)
    return 0;

if __name__ == "__main__":
    main()

File: cpp/day_19b.cpp
#include <algorithm>
#include <cassert>
#include <fstream>
#include <functional>
#include <iostream>
#include <map>
#include <string>
#include <vector>

// No need to change the input data
const std::map<int, std::string> update_rules{{8, "8: 42 | 42 8"},
                                              {11, "11: 42 31 | 42 11 31"}};

int main() {
  std::fstream file{"../input/day_19_input"};
  std::string s;
  std::map<int, std::string> known_rules;
  std::map<int, std::vector<std::vector<int>>> subsections_of_rules;
  std::vector<std::string> messages;
  // TODO(vss): improve the way line is read
  while (std::getline(file, s)) {
    s.erase(std::remove_if(std::begin(s), std::end(s),
                           [](const char c) { return !isprint(c); }),
            std::end(s));
    if (s.find(":") != std::string::npos) {
      int rule_lhs = std::stoi(s.substr(0, s.find(":")));
      if (const auto it = update_rules.find(rule_lhs);
          it != update_rules.end()) {
        s = it->second;
      }
      if (s.find("\"") != std::string::npos) {
        std::string suffix = s.substr(s.find("\"") + 1, 1);
        known_rules[rule_lhs] = suffix;
      } else {
        s = s.substr(s.find(":") + 2);
        std::vector<int> rule_section;
        int num_in_rule_section = 0;
        for (char c : s) {
          if (c == ' ') {
            if (num_in_rule_section > 0) {
              rule_section.push_back(num_in_rule_section);
            }
            num_in_rule_section = 0;
          } else if (c == '|') {
            subsections_of_rules[rule_lhs].push_back(rule_section);
            rule_section.clear();
          } else {
            num_in_rule_section = (num_in_rule_section * 10) + (c - '0');
          }
        }
        rule_section.push_back(num_in_rule_section);
        subsections_of_rules[rule_lhs].push_back(rule_section);
      }
    } else {
      messages.push_back(s);
    }
  }
  std::map<std::pair<std::string, int>, bool> previously_seen_pairs;
  std::function<bool(std::string, int)> DoesMsgFollowRule =
      [&previously_seen_pairs, &known_rules, &subsections_of_rules,
       &DoesMsgFollowRule](const std::string& s, const int source) {
        const int n = s.size();
        // If the rule is a known rule, return whether the string matches the
        // known rule
        if (known_rules.count(source)) {
          return (s == known_rules[source]);
        }
        // If a particular {pattern, rule index} pair has already been seen
        // before, retuurn the stored result
        if (previously_seen_pairs.count({s, source})) {
          return previously_seen_pairs[{s, source}];
        }
        // Iterate over each subsection to check whether the string satisfies
        // any of the subsections
        for (auto& subsection : subsections_of_rules[source]) {
          const size_t parts = subsection.size();
          if (parts == 1) {
            // If 1 number in subsection, check whether the string satisfies the
            // corresponding rule
            if (DoesMsgFollowRule(s, subsection[0])) {
              return previously_seen_pairs[{s, source}] = true;
            }
          } else if (parts == 2) {
            // If 2 numbers in subsection, check whether the string can be
            // broken in a way to satisfy both numbers in the subsection
            // simultaneously
            for (int i = 0; i < n - 1; i++) {
              if (DoesMsgFollowRule(s.substr(0, i + 1), subsection[0]) &&
                  DoesMsgFollowRule(s.substr(i + 1), subsection[1])) {
                return previously_seen_pairs[{s, source}] = true;
              }
            }
          } else if (parts == 3) {
            // If 3 numbers in subsection, check whether the string can be
            // broken in a way to satisfy all 3 numbers in the subsection
            // simultaneously
            for (int i = 0; i < n; i++) {
              for (int j = i + 1; j < n; j++) {
                if (DoesMsgFollowRule(s.substr(0, i + 1), subsection[0]) &&
                    DoesMsgFollowRule(s.substr(i + 1, j - i), subsection[1]) &&
                    DoesMsgFollowRule(s.substr(j + 1, n - j), subsection[2])) {
                  return previously_seen_pairs[{s, source}] = true;
                }
              }
            }
          } else {
            std::cout << "This should not happen" << '\n';
          }
        }
        return previously_seen_pairs[{s, source}] = false;
      };
  int res = 0;
  for (const std::string& message : messages) {
    if (DoesMsgFollowRule(message, 0)) {
      ++res;
    }
    previously_seen_pairs.clear();
  }
  std::cout << res << '\n';
  return 0;
}
